Assets/HoloToolkit/Input/README.md

```diff
@@ -3,50 +3,649 @@
+        Debug.LogFormat("OnManipulationCompleted\r\nSource: {0}  SourceId: {1}\r\nCumulativeDelta: {2} {3} {4}",
+            eventData.InputSource,
+            eventData.SourceId,
+            eventData.CumulativeDelta.x,
+            eventData.CumulativeDelta.y,
+            eventData.CumulativeDelta.z);
+
+        eventData.Use(); // Mark the event as used, so it doesn't fall through to other handlers.
+    }
+
+    void IManipulationHandler.OnManipulationCanceled(ManipulationEventData eventData)
+    {
+        Debug.LogFormat("OnManipulationCanceled\r\nSource: {0}  SourceId: {1}\r\nCumulativeDelta: {2} {3} {4}",
+            eventData.InputSource,
+            eventData.SourceId,
+            eventData.CumulativeDelta.x,
+            eventData.CumulativeDelta.y,
+            eventData.CumulativeDelta.z);
+
+        eventData.Use(); // Mark the event as used, so it doesn't fall through to other handlers.
+    }
+
+    void INavigationHandler.OnNavigationStarted(NavigationEventData eventData)
+    {
+        Debug.LogFormat("OnNavigationStarted\r\nSource: {0}  SourceId: {1}\r\nCumulativeDelta: {2} {3} {4}",
+            eventData.InputSource,
+            eventData.SourceId,
+            eventData.NormalizedOffset.x,
+            eventData.NormalizedOffset.y,
+            eventData.NormalizedOffset.z);
+
+        eventData.Use(); // Mark the event as used, so it doesn't fall through to other handlers.
+    }
+
+    void INavigationHandler.OnNavigationUpdated(NavigationEventData eventData)
+    {
+        Debug.LogFormat("OnNavigationUpdated\r\nSource: {0}  SourceId: {1}\r\nCumulativeDelta: {2} {3} {4}",
+            eventData.InputSource,
+            eventData.SourceId,
+            eventData.NormalizedOffset.x,
+            eventData.NormalizedOffset.y,
+            eventData.NormalizedOffset.z);
+
+        eventData.Use(); // Mark the event as used, so it doesn't fall through to other handlers.
+    }
+
+    void INavigationHandler.OnNavigationCompleted(NavigationEventData eventData)
+    {
+        Debug.LogFormat("OnNavigationCompleted\r\nSource: {0}  SourceId: {1}\r\nCumulativeDelta: {2} {3} {4}",
+            eventData.InputSource,
+            eventData.SourceId,
+            eventData.NormalizedOffset.x,
+            eventData.NormalizedOffset.y,
+            eventData.NormalizedOffset.z);
+
+        eventData.Use(); // Mark the event as used, so it doesn't fall through to other handlers.
+    }
+
+    void INavigationHandler.OnNavigationCanceled(NavigationEventData eventData)
+    {
+        Debug.LogFormat("OnNavigationCanceled\r\nSource: {0}  SourceId: {1}\r\nCumulativeDelta: {2} {3} {4}",
+            eventData.InputSource,
+            eventData.SourceId,
+            eventData.NormalizedOffset.x,
+            eventData.NormalizedOffset.y,
+            eventData.NormalizedOffset.z);
+
+        eventData.Use(); // Mark the event as used, so it doesn't fall through to other handlers.
+    }
+}
+```
+
+### Motion Controller Input
```

Comment:
If I were using this, one question I'd have is: "how do I access the various buttons that are on a 6DOF controller, such as grasp, thumbstrick, etc."?

We'll need to figure out how we want to handle these, which certainly shouldn't happen in this CR, but I'm hijacking your review as a way to communicate (sorry!) :) . In some internal prototypes, the "first idea that works" implementation we did, we essentially added the following interfaces:

IControllerGraspHandler.cs
IControllerMenuHandler.cs
IControllerSelectTriggerHandler.cs
IControllerThumbstickHandler.cs
IControllerTouchpadHandler.cs

This gives you an interface for each standardized button that Windows MR 6DOF controllers can support. I initially wanted to have a more generic button handling event, but realized that this would mean that components would have to generically handle all buttons , which would lead to more expensive event bubbling. I guess it's a perf compromise we'll need to figure out later.

Reply:
I'm pretty sure these interfaces are already there, are they not?

Reply:
I don't think so. At least I can't find them under InputHandlers.

Also even the one that's there for Touchpad is incomplete. It only includes touched/released, but no position and no pressed/released. Mine looked like this instead:

    public interface IControllerTouchpadHandler : IEventSystemHandler
    {
        void OnTouchpadTouched(InputEventData eventData);
        void OnTouchpadUntouched(InputEventData eventData);
        void OnTouchpadReleased(InputEventData eventData);
        void OnTouchpadPressed(InputEventData eventData);
        void OnTouchpadXYChanged(InputXYEventData eventData);
    }

I'll see if I can find some time before the end of the year to get it in somehow...

Reply:
Could we make sure they're `MotionController` or something that doesn't sound as generic?

Reply:
All of that exists under `InputHandlers`. I'm sure there are improvements to the names or event groupings that could be made though.

Touchpad XY comes from [IControllerInputHandler.cs(]https://github.com/Microsoft/MixedRealityToolkit-Unity/blob/master/Assets/HoloToolkit/Input/Scripts/InputHandlers/IControllerInputHandler.cs)

Touchpad Pressed and Released comes from [IInputHandler.cs](https://github.com/Microsoft/MixedRealityToolkit-Unity/blob/master/Assets/HoloToolkit/Input/Scripts/InputHandlers/IInputHandler.cs)

Reply:
Ah I see. That does provide the functionality, although I can certainly see how it could be hard to figure out that touched/untouched is in a different interface than pressed/released. 

From an input routing perspective, I also wonder what the perf impact is. It's probably acceptable, but because you have a single interface, you essentially end up potentially a lot more code running every time a button is clicked, because more handlers are getting called when a button is pressed (even if most of them end up not doing anything because they don't care about the specific button that was pressed).

It's a tricky problem though, so I guess that implementation is fine for now. I do think we should better document how to get all the various input events for the 6DOF controller buttons though, because it's not super obvious right now.

Reply:
@maxouellet I'd like to make events a little more generic.  Like having a type we can select from and subscribe to the type raised.  That way we don't have to have a bunch of different interfaces for each of these events that are pretty much the same.

Something like:
`IInputButtonPress`
`IInputButtonDown`
`IInputButtonUp`
`IInputTouchPress`
`IInputTouchDown`
`IInputTouchUp`
`IInputAxisChange`
`IInputPositionChanged`

Reply:
Yeah that would be one way to go about it. The downside is that it will be less performant. As an example, anything that cares about a button press would need to implement the generic IInputButtonPress, so all button presses would be sent to that script, even if the script only cares about one button in particular. And all scripts would have to implement something like:

if (eventData.button == BUTTON.WHATEVER)
{
   ...
   eventData.handled = true;
}

I'm definitely not against it if that's what the community wants though.

Reply:
Can you give an example of how it would impact performance?

I think we would have to really think about how to send this type of information through.
I could see it using the system that @Railboy and @KatVHarris are doing for Multi Pointers.

Reply:
I did give an example. You have more listeners that don't care about the events they receive, thus the events need to bubble up more. I'm not saying it would create massive perf issues, but it would end up needing to make more function calls, thus be a little more expensive in some scenarios. Doesn't necessarily mean it can't be done, it might be an unnoticeable perf hit

Reply:
If it's an enum check, I don't see it being too crazy.

Reply:
I think the Multi Pointer system they're working on might end up doing something very similar.
I'd like you to check that out if you get time.